kernel/kernel-api/src/main/java/io/delta/kernel/internal/transaction/TransactionDataSource.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (2025) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.kernel.internal.transaction;
+
+import io.delta.kernel.internal.actions.Metadata;
+import io.delta.kernel.internal.actions.Protocol;
+
+/**
+ * As we transition from {@link io.delta.kernel.Transaction} (and {@link io.delta.kernel.Snapshot})
+ * to {@link io.delta.kernel.transaction.TransactionV2} (and {@link io.delta.kernel.ResolvedTable}),
+ * different parts of internal Kernel code will have to deal with both types of table "sources"
+ * (Snapshot, ResolvedTable).
+ *
+ * <p>This is a common interface for the minimal set of methods that both sources will have to
+ * expose.
+ */
+public interface TransactionDataSource {
```

Comment:
So does this _only_ exist so we can have _either_ a `Snapshot` _or_ a `ResolvedTable` back the txn? 

I'm kinda okay with that, but I do think we'll probably want `Snapshot` to go away at some point and just become `ResolvedTable`. That's not fully settled yet. 

Is it hard to have `Snapshot` implement `ResolvedTable`? Looking at the `ResolvedTabled` APIs, I think it should be quite simple. Then a txn could just have a `ResolvedTable` that backs it and we can get rid of this interface.

Reply:
> So does this only exist so we can have either a Snapshot or a ResolvedTable back the txn?

Yes. We have lots of helper methods today (e.g. in TransactionImpl.java, in ConflictChecker.java, for Row Tracking, for Domain Metadata, etc.)

We will have to update all of those APIs to take in _something_ common to both ResolvedTable and Snapshot, since we have to continue to support the Snapshot public API for now.

I think `interface TransactionDataSource` is a good way to go about that 👍 

> Is it hard to have Snapshot implement ResolvedTable?

I don't think we want to change our public Snapshot API? It would have to be able to produce a Committer and a TransactionBuilder

Reply:
Mmm okay. The point is that we wouldn't have to update those APIs at all when we deprecated snapshot if they just took a `ResolvedTable`. Assumably nowhere that we'd use a snapshot like that would we actually try and get a committer or builder. So we could have the implementation for those just throw, but I agree it's a bit awkward.

I'll leave it up to your judgement here.